ibm/resource_ibm_cr_retention_policy.go

```diff
@@ -0,0 +1,183 @@
+/**
+ * (C) Copyright IBM Corp. 2021.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ibm
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/IBM/container-registry-go-sdk/containerregistryv1"
+)
+
+func resourceIBMCrRetentionPolicy() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceIBMCrRetentionPolicyCreate,
+		ReadContext:   resourceIBMCrRetentionPolicyRead,
+		UpdateContext: resourceIBMCrRetentionPolicyUpdate,
+		DeleteContext: resourceIBMCrRetentionPolicyDelete,
+		Importer:      &schema.ResourceImporter{},
+
+		Schema: map[string]*schema.Schema{
+			"namespace": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The namespace to which the retention policy is attached.",
+			},
+			"images_per_repo": &schema.Schema{
+				Type:        schema.TypeInt,
+				Optional:    true,
```

Comment:
Based on this comment
// A retention policy "does not exist" if `imagesPerRepo` and `retainUntagged` have their default values

Since you are definign default values  for images_per_repo and retain_untagged then if user provides
resource "ibm_cr_retention_policy" "cr_retention_policy" {
			namespace = ibm_cr_namespace.cr_namespace.name
		}
		
		Then will API throws back an error imagesPerRepo and retainUntagged both are set to default vlaues?
		or can we add condition check and throws back an error befor calling the create retention policy?

Reply:
Thinking...

Reply:
Ok - firstly, I will correct that comment... a retention policy does not exist if `imagesPerRepo` is `-1` and `retainUntagged` is `true` , the default value for `retainUntagged` when creating or updating a policy is actually `false`.
However, I think the code is mostly accurate in this regard.

I agree it would make sense, in the context of creating a retention policy, to throw back an error if images_per_repo is not set. The equivalent CLI command for this treats this parameter as mandatory. 
I'll fix this up.

Reply:
Believe I have addressed this on line 58-60:
```Go
} else {
		return diag.Errorf("images_per_repo must be set when creating a retention policy")
	}
```

Reply:
Hmm - maybe this isn't right. In local testing, it looks like Terraform just uses the default value in this case (-1), and the `else` block doesn't get hit.
@hkantare - Do you have any guidance here? Should we check for it being set to -1 here, instead of (or as well as) checking for it not being set at all?

Reply:
ok - I've refixed this as outlined in my previous comment. This now works when tested locally.

Reply:
```
I agree it would make sense, in the context of creating a retention policy, to throw back an error if images_per_repo is not set. The equivalent CLI command for this treats this parameter as mandatory.
I'll fix this up.
```

Then can terraform also make  images_per_repo as required because still i'm in confusion if user writes a template
```
resource "ibm_cr_retention_policy" "cr_retention_policy" {
namespace = ibm_cr_namespace.cr_namespace.name
}
```

Then by default images_per_repo is -1 and terraform apply fails
````
images_per_repo must be set when creating a retention policy
```
That implies that images_per_repo is required parameter

Reply:
@hkantare - yes, we could go down that route... 
Strictly speaking, you might want to update a retention policy to change the retain_untagged behaviour, without modifying images_per_repo.. The API allows this, but our CLI won't currently let you do it - so probably reasonable just to make it a required parameter. I'll make the change, and let you know when it's ready for re-review.

To answer your question directly - if you were to create this policy (if our provider allowed it):
```
resource "ibm_cr_retention_policy" "cr_retention_policy" {
namespace = ibm_cr_namespace.cr_namespace.name
}
```

You would end up with all images retained, but the retain_untagged would flip from true to false.. This is also a bit esoteric and not easy for users to understand - which strengthens the argument for just making images_per_repo mandatory.